packages/unified-latex-to-pretext/libs/pre-conversion-subs/break-on-boundaries.ts

```diff
@@ -0,0 +1,134 @@
+import { env, arg } from "@unified-latex/unified-latex-builder";
+import * as Ast from "@unified-latex/unified-latex-types";
+import { getNamedArgsContent } from "@unified-latex/unified-latex-util-arguments";
+import {
+    anyEnvironment,
+    anyMacro,
+    match,
+} from "@unified-latex/unified-latex-util-match";
+import { printRaw } from "@unified-latex/unified-latex-util-print-raw";
+import { replaceNode } from "@unified-latex/unified-latex-util-replace";
+import {
+    splitOnMacro,
+    unsplitOnMacro,
+} from "@unified-latex/unified-latex-util-split";
+import { visit } from "@unified-latex/unified-latex-util-visit";
+
+// all divisions [division macro, environment]
+const divisions: [string, string][] = [
+    ["part", "_part"],
+    ["chapter", "_chapter"],
+    ["section", "_section"],
+    ["subsection", "_subsection"],
+    ["subsubsection", "_subsubsection"],
+    ["paragraph", "_paragraph"],
+    ["subparagraph", "_subparagraph"],
+];
+
+/**
+ * Breaks up division macros into environments. Returns a list of warning messages
+ * for any groups that were removed.
+ */
+export function breakOnBoundaries(ast: Ast.Ast): string[] {
+    // messages for any groups removed
+    const messages: string[] = [];
+
+    visit(ast, (node, info) => {
+        // needs to be an environment, root, or group node
+        if (
+            !(
+                anyEnvironment(node) ||
+                node.type === "root" ||
+                match.group(node)
+            ) ||
+            // skip math mode
+            info.context.hasMathModeAncestor
+        ) {
+            return;
+        }
+        // if it's an environment, make sure it isn't a newly created one
+        else if (
+            anyEnvironment(node) &&
+            divisions.map((x) => x[1]).includes(node.env)
+        ) {
+            return;
+        }
+
+        // if it's a group, push a warning message
+        if (match.group(node)) {
+            messages.push(
+                `Warning: hoisted out of a group, which might break the LaTeX code. { group: ${printRaw(
+                    node
+                )} }`
+            );
+        }
+
+        // now break up the divisions, starting at part
+        node.content = breakUp(node.content, divisions, 0);
+    });
+
+    replaceNode(ast, (node) => {
+        // remove all old division nodes
+        if (
+            anyMacro(node) &&
+            divisions.map((x) => x[0]).includes(node.content)
```

Comment:
This is very inefficient since it is being rebuilt for every single node. Make a custom macro matcher before the call to `replaceNode` and use that instead.